### PR TITLE
Create User - creation of gnosisSafe proxy, tokenHolder proxy, recovery proxy

### PR DESCRIPTION
**Pull request by abhayks1**
The PR covers implementation of below action items:

- Creation of user wallet i.e. creates gnosisSafe proxy, tokenHolder proxy, recovery proxy.
- Readme update.
- Bug fix while printing UBT address for an economy.

Command:
```
node ./src/bin/create_user.js config.json utilityBrandedTokenAddress owners threshold sessionKeys sessionKeySpendingLimits sessionKeyExpirationHeights
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -229,6 +229,37 @@ It will write `tokenRulesAddress` to your config file.
...
+  ```
+
+  * Replace `config.json` with the path to the configuration file.
+  * Replace `eip20Token` with the UBT address of the economy.
+  * Replace `owners` with comma separated owner addresses.
+  * Replace `threshold` with gnosis requirement. 
+  * Replace `sessionKeys` comma separated ephemeral keys. 
+  * Replace `sessionKeySpendingLimits` comma separated spending limits corresponding to ephemeral keys. 
+  * Replace `sessionKeyExpirationHeights` comma separated expiration heights corresponding to ephemeral keys. 
+
```
Comment (schemar) [inline README.md]: why spaces in beginning of lines?

File: src/bin/create_user.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      sessionKeys,
+      sessionKeySpendingLimits,
+      sessionKeyExpirationHeights
+    ) => {
+      const chainConfig = new ChainConfig(config);
+      const createUserTxOptions = {
+        from: chainConfig.auxiliaryMasterKey,
+        gasPrice: chainConfig.auxiliaryGasPrice,
+      };
+      const auxiliaryWeb3 = new Web3(chainConfig.auxiliaryWeb3Provider);
```
Comment (schemar) [inline src/bin/create_user.js]: use `connected::run()` instead.

File: src/config/chain_config.js [hunk 3]
```diff
@@ -7,8 +7,8 @@ class ChainConfig {
    * @param {string} [filePath] An optional config file to overwrite any defaults or env variables.
    */
   constructor(filePath) {
-    this.originWeb3Provider = process.env.ORIGIN_WEB3_PROVIDER || 'http://localhost:8545';
-    this.auxiliaryWeb3Provider = process.env.AUXILIARY_WEB3_PROVIDER || 'http://localhost:8547';
+    this.originWeb3Provider = process.env.ORIGIN_WEB3_PROVIDER || 'http://localhost:8546';
+    this.auxiliaryWeb3Provider = process.env.AUXILIARY_WEB3_PROVIDER || 'http://localhost:8545';
```
Comment (schemar) [inline src/config/chain_config.js]: why change defaults?

File: README.md (hunk 1 above)
Comment (abhayks1) [reply README.md]: Remove the spaces. it's not needed 👍

File: src/bin/create_user.js (hunk 2 above)
Comment (abhayks1) [reply src/bin/create_user.js]: Done!

**Comment by abhayks1** [comment]
@schemar Changes has been done. Will test together with openst setup PR.

File: src/config/chain_config.js (hunk 3 above)
Comment (abhayks1) [reply src/config/chain_config.js]: Fixed 👍  Defaults doesn't needs to be changed.

File: README.md [hunk 4]
```diff
@@ -233,6 +233,37 @@ It will write `tokenRulesAddress` to your config file.
...
+  
+# Create User:
+node ./src/bin/create_user.js <config.json> <eip20Token> <owners> <threshold> <sessionKeys> <sessionKeySpendingLimits> <sessionKeyExpirationHeights>
+```
+
+* Replace `config.json` with the path to the configuration file.
+* Replace `eip20Token` with the UBT address of the economy.
+* Replace `owners` with comma separated owner addresses.
+* Replace `threshold` with gnosis requirement. 
+* Replace `sessionKeys` comma separated ephemeral keys.
```
Comment (sarvesh-ost) [inline README.md]:
```suggestion
* Replace `sessionKeys` comma separated session keys. 
```

Comment (sarvesh-ost) [reply README.md]: Do we use still use ephemeral key term? If not please fix at other places.

File: src/bin/create_user.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    '--help',
+    () => {
+      console.log('');
+      console.log('Arguments:');
+      console.log('  config                       path to a config file.');
+      console.log('  eip20Token                   EIP20Token address of an economy.');
+      console.log('  owners                       comma separated owners.');
+      console.log('  threshold                    gnosis requirement.');
+      console.log('  sessionKeys                  comma separated ephemeral keys.');
+      console.log('  sessionKeySpendingLimits    comma separated session keys spending limits.');
```
Comment (sarvesh-ost) [inline src/bin/create_user.js]:
```suggestion
      console.log('  sessionKeySpendingLimits     comma separated session keys spending limits.');
```

File: src/bin/create_user.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env node
+
+'use strict';
+
+const program = require('commander');
+const Web3 = require('web3');
+const OpenST = require('@openstfoundation/openst.js');
```
Comment (sarvesh-ost) [inline src/bin/create_user.js]:
```suggestion
const {GnosisSafe, Helpers} = require('@openstfoundation/openst.js');
```

File: src/bin/create_user.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            eip20Token,
+            chainConfig.openst.tokenRules,
+            chainConfig.openst.userWalletFactory,
+            chainConfig.openst.proxyFactory,
+            auxiliaryWeb3,
+          );
+          const ownersArray = owners.split(',').map(item => item.trim());
+          const sessionKeysArray = sessionKeys.split(',').map(item => item.trim());
+          const sessionKeySpendingLimitsArray = sessionKeySpendingLimits.split(',').map(item => item.trim());
+          const sessionKeyExpirationHeightsArray = sessionKeyExpirationHeights.split(',').map(item => item.trim());
```
Comment (sarvesh-ost) [inline src/bin/create_user.js]: I think it's good to add a validation that all array lengths are the same.

File: src/bin/create_user.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            chainConfig.openst.recoveryOwnerAddress,
+            chainConfig.openst.recoveryControllerAddress,
+            chainConfig.openst.recoveryBlockDelay,
+            sessionKeysArray,
+            sessionKeySpendingLimitsArray,
+            sessionKeyExpirationHeightsArray,
+            createUserTxOptions,
+          );
+          const { returnValues } = response.events.UserWalletCreated;
+          const gnosisSafeProxy = returnValues._gnosisSafeProxy;
```
Comment (sarvesh-ost) [inline src/bin/create_user.js]: Do you not want to write these value back in the config?

File: README.md (hunk 4 above)
Comment (abhayks1) [reply README.md]: makes sense 👍

File: src/bin/create_user.js (hunk 7 above)
Comment (abhayks1) [reply src/bin/create_user.js]: Agree 👍
But I  think it should be part of createUserWallet interaction method.

File: src/bin/create_user.js (hunk 8 above)
Comment (abhayks1) [reply src/bin/create_user.js]: Good Point 💯 Will update!

**Pull request merged**